test/cppsuite/tests/background_compact.cpp

```diff
@@ -0,0 +1,328 @@
+
+        /* Loop while the test is running. */
+        while (tw->running()) {
+            /* Make sure we're not doing any work during the maintenance window. */
+            if (maintenance_window) {
+                std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+                continue;
+            }
+
+            /*
+             * Sleep the period defined by the op_rate in the configuration. Do this at the start of
+             * the loop as it could be skipped by a subsequent continue call.
+             */
+            tw->sleep();
+
+            /* Choose a random collection to truncate. */
+            collection &coll = tw->db.get_random_collection();
+
+            /* Look for an existing random cursor in our cursor cache. */
+            if (rnd_cursors.find(coll.id) == rnd_cursors.end()) {
+                logger::log_msg(
+                  LOG_TRACE, log_prefix + "Creating cursors for collection: " + coll.name);
+                /* Open the two cursors for the chosen collection. */
+                scoped_cursor rnd_cursor =
+                  tw->session.open_scoped_cursor(coll.name, "next_random=true");
+                rnd_cursors.emplace(coll.id, std::move(rnd_cursor));
+                std::string stat_uri = STATISTICS_URI + coll.name;
+                scoped_cursor stat_cursor = tw->session.open_scoped_cursor(stat_uri);
+                stat_cursors.emplace(coll.id, std::move(stat_cursor));
+            }
+
+            /* Get the cursors associated with the collection. */
+            scoped_cursor &stat_cursor = stat_cursors[coll.id];
+            scoped_cursor &rnd_cursor = rnd_cursors[coll.id];
+
+            /* Get the file statistics so we know how much to truncate. */
+            int64_t entries, bytes_avail_reuse, file_size;
+            metrics_monitor::get_stat(stat_cursor, WT_STAT_DSRC_BTREE_ENTRIES, &entries);
+            metrics_monitor::get_stat(
+              stat_cursor, WT_STAT_DSRC_BLOCK_REUSE_BYTES, &bytes_avail_reuse);
+            metrics_monitor::get_stat(stat_cursor, WT_STAT_DSRC_BLOCK_SIZE, &file_size);
+
+            /* Don't truncate if we already have enough free space for compact to do work. */
+            int64_t pct_free_space = (bytes_avail_reuse * 100 / file_size);
+            if (pct_free_space > 10) {
+                logger::log_msg(LOG_INFO,
+                  log_prefix + "Skip truncating coll " + std::to_string(coll.id) +
+                    " free space available = " + std::to_string(pct_free_space));
+                testutil_check(stat_cursor->reset(stat_cursor.get()));
+                continue;
+            }
+
+            /*
+             * Truncate a range of keys between 0 and 100 until we've truncated a total of 20% of
+             * the entries in the table.
+             */
+            int64_t n_keys_to_truncate = (entries / 100) * 20;
+            int64_t keys_truncated = 0;
+            while (keys_truncated < n_keys_to_truncate) {
```

Comment:
Please check the thread is supposed to keep running here too with `tw->running()`. Furthermore, I am afraid we could be stick in an infinite loop if we never manage to truncate that much for some reason. We could have N max retries.

Reply:
Good point, yes that could be possible, I'll add a max retries check.

Reply:
Added here https://github.com/wiredtiger/wiredtiger/blob/0014dfa45ae7c27ab3a958f59f8131232b05ea77/test/cppsuite/tests/background_compact.cpp#L136-L138